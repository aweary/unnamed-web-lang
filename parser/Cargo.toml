[package]
name = "parser"
version = "0.1.0"
authors = ["Brandon Dail <brandondail@fb.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
graphql-parser = "0.2.3"
salsa = "0.14.1"
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
diagnostics = { path = "../diagnostics" }
source = { path = "../source" }
insta = { version = "0.16", features = ["glob"] }
log = "0.4.8"