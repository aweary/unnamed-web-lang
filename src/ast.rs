#[derive(Debug)]
pub struct Param {
    name: String,
    kind: String,
}

impl Param {
    pub fn new(name: String, kind: String) -> Self {
        Param { name, kind }
    }
}

/**
 * The definition of a component. Similar to a function definition, but with
 * stricter semantics and requirements.
 *
 * @example
 * component Button(label: string) {
 *   <button>{label}</button>
 * }
 *
 */
#[derive(Debug)]
pub struct ComponentDefinition {
    // Name of the component
    name: String,
    // List of its parameters, which must by typed
    params: Option<Vec<Param>>,
    // The content of the component
    body: Vec<Expression>,
}

impl ComponentDefinition {
    pub fn new(name: String, params: Option<Vec<Param>>, body: Vec<Expression>) -> Self {
        ComponentDefinition { name, params, body }
    }
}

#[derive(Debug)]
pub struct TypeDefinition {
    name: String,
}

impl TypeDefinition {
    pub fn new(name: String) -> Self {
        TypeDefinition { name }
    }
}

#[derive(Debug)]
pub struct Identifier {
    name: String,
}

impl Identifier {
    pub fn new(name: String) -> Self {
        Identifier { name }
    }
}

#[derive(Debug)]
pub struct JSXOpeningElement {
    tag_name: String,
    attributes: Option<Vec<JSXAttribute>>,
    self_closing: bool,
}

impl JSXOpeningElement {
    pub fn new(
        tag_name: String,
        attributes: Option<Vec<JSXAttribute>>,
        self_closing: bool,
    ) -> Self {
        JSXOpeningElement {
            tag_name,
            attributes,
            self_closing,
        }
    }
}

#[derive(Debug)]
pub struct JSXClosingElement {
    tag_name: String,
}

impl JSXClosingElement {
    pub fn new(tag_name: String) -> Self {
        JSXClosingElement { tag_name }
    }
}

#[derive(Debug)]
pub struct JSXAttribute {
    name: String,
    value: JSXAttributeValue,
}

impl JSXAttribute {
    pub fn new(name: String, value: JSXAttributeValue) -> Self {
        JSXAttribute { name, value }
    }
}

#[derive(Debug)]
pub struct JSXElement {
    opening_element: JSXOpeningElement,
    closing_element: Option<JSXClosingElement>,
    children: Option<Vec<Box<JSXChildren>>>,
}

impl JSXElement {
    pub fn new(
        opening_element: JSXOpeningElement,
        closing_element: Option<JSXClosingElement>,
        children: Option<Vec<Box<JSXChildren>>>,
    ) -> Self {
        JSXElement {
            opening_element,
            closing_element,
            children,
        }
    }
}

#[derive(Debug)]
pub enum Expression {
    NumericLiteral(i64),
    StringLiteral(String),
    Identifer(Identifier),
    JSXElement(JSXElement),
}

#[derive(Debug)]
pub enum JSXAttributeValue {
    StringLiteral(String),
    Expression(Expression),
}

#[derive(Debug)]
pub enum JSXChildren {
    JSXText(String),
    JSXExpressionContainer(Expression),
    JSXElement(JSXElement),
}

// pub enum Expression {
//     JSXElement,
//     JSXOpeningElement,
//     JSXClosingElement,
//     JSXAttribute,
// }

#[derive(Debug)]
pub enum Node {
    ComponentDefinition(ComponentDefinition),
    TypeDefinition(TypeDefinition),
    // Identifier,
    // OpeningElement,
    // ClosingElement,
    Text,
    // ExpressionStatement,
    // StateCellDeclaration,
}
