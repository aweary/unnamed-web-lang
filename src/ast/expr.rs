use crate::symbol::Symbol;
use crate::typecheck::TKind;
use id_arena::{Id};


#[derive(Debug)]
pub struct Expr {
    pub kind: Box<ExprKind>,
    /**
     * The internal type of the expression. For some expression kinds
     * we can determine this at parse time (specifically, literals like
     * numbers, strings, booleans, etc.)
     */
    pub ty: Option<TKind>,
}

pub type ExprId = Id<Expr>;

impl Expr {
    pub fn new(kind: ExprKind) -> Self {
        Expr {
            kind: Box::new(kind),
            ty: None,
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum Op {
    Add,
    Sub,
    Div,
    Mul,
    Mod,
    And,
    Or,
    GreaterThan,
    LessThan,
}

#[derive(Debug)]
pub struct MatchArm {
    pub test: ExprId,
    pub consequent: ExprId,
}

#[derive(Debug)]
pub struct ObjectProperty(pub Symbol, pub ExprId);

#[derive(Debug)]
pub struct Call {
    pub callee: ExprId,
    pub arguments: Vec<ExprId>,
}

#[derive(Debug)]
pub enum ExprKind {
    Number(f64),
    Str(Symbol),
    Ident(Symbol),
    Bool(bool),
    Unary(Op, ExprId),
    Array(Vec<ExprId>),
    Object(Vec<ObjectProperty>),
    Binary {
        op: Op,
        left: ExprId,
        right: ExprId,
    },
    Cond {
        test: ExprId,
        consequent: ExprId,
        alternate: ExprId,
    },
    Logical {
        op: Op,
        left: ExprId,
        right: ExprId,
    },
    Call(Call),
    Member {
        obj: ExprId,
        property: Symbol,
    },
    Match {
        discriminant: ExprId,
        cases: Vec<MatchArm>,
    },
}

pub fn number_expr(num: f64) -> Expr {
    let mut expr = Expr::new(
        ExprKind::Number(num)
    );
    expr.ty = Some(TKind::Number);
    expr
}