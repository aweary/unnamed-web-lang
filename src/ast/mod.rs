mod expr;
mod stmt;

use crate::symbol::Symbol;
use crate::typecheck::TKind;

pub use expr::*;
pub use stmt::*;


// The root node of the AST.
#[derive(Debug)]
pub struct Program {
    // The set of all modules the program requires.
    pub modules: Vec<Module>,
    // Arena that stores all the template references.
}

// An individual module. Currently the only way to define
// a module is with a new file, so this maps directly to the filesystem
// layout of the project.
#[derive(Debug)]
pub struct Module {
    pub stmts: Vec<Decl>,
}

#[derive(Debug, PartialOrd, Ord, PartialEq, Eq, Clone, Copy)]
pub enum Precedence {
    NONE = 0,
    ASSIGNMENT = 1,
    CONDITIONAL = 2,
    SUM = 3,
    PRODUCT = 4,
    COMPARE = 5,
    PREFIX = 6,
}






