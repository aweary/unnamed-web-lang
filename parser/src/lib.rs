#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::must_use_candidate)]

mod parser;

pub use crate::parser::*;

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use insta::assert_json_snapshot;
    use syntax::ast;

    fn parse_module(source: &str) -> ast::Module {
        let mut parser = Parser::new(source);
        parser.parse_module().unwrap()
    }

    #[test]
    fn fixture_function_simple() {
        let input = include_str!("./inputs/function_simple.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_control_flow() {
        let input = include_str!("./inputs/control_flow.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_trailing_closure_control_flow() {
        let input = include_str!("./inputs/trailing_closures.dan");
        assert_json_snapshot!(parse_module(input));
    }
}
