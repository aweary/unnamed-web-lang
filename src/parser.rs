use ast::*;
use lexer::Lexer;
use token::*;

pub struct Parser {
    pub lexer: Lexer,
    // state: ParserState,
}

impl Parser {
    pub fn module(source: String) -> Self {
        let lexer = Lexer::new(source);
        // let state = ParserState::new();
        Parser { lexer }
    }

    fn peek(&mut self) -> Option<&Token> {
        self.lexer.peek_token()
    }

    pub fn skip_until<F>(&mut self, predicate: F)
    where
        F: Fn(Token) -> bool,
    {
        loop {
            match self.lexer.next_token() {
                Some(token) => {
                    if predicate(token) {
                        return;
                    }
                }
                None => return,
            }
        }
    }

    pub fn skip_assert(&mut self, token: Token) {
        assert_eq!(Some(token), self.lexer.next_token())
    }

    fn read_param_list(&mut self) -> Option<Vec<Param>> {
        match self.lexer.next_token() {
            // If the next token is an opening curly bracket, there
            // are no arguments and we assume this component takes no input
            Some(Token::LCurlyBracket) => None,
            // If the next token is an opening paren, there is
            // an argument list.
            Some(Token::LParen) => {
                let mut params: Vec<Param> = vec![];
                loop {
                    let name = self.read_ident();
                    // TODO this should be an unexpected token error
                    assert_eq!(Some(Token::Colon), self.lexer.next_token());
                    let kind = self.read_ident();
                    params.push(Param::new(name, kind));
                    match self.lexer.next_token() {
                        Some(Token::RParen) => {
                            break;
                            // End of the param list
                        }
                        Some(Token::Comma) => {
                            continue;
                            // There's another paramter
                        }
                        // Anything else is an unexpected token error
                        _ => unreachable!(),
                    }
                }
                Some(params)
            }
            _ => unreachable!("Unexpected token"),
        }
        // Next token should be the opening paren for the argument list
    }

    fn read_ident(&mut self) -> String {
        match self.lexer.next_token() {
            Some(Token::Ident(name)) => name,
            unknown @ _ => unreachable!("Expected identifier, found: {:?}", unknown),
        }
    }

    // Assumes that the current token is Reserved(Component)
    fn read_component_declaration(&mut self) -> Node {
        // Next token should be the identifier for the component definition
        let name = self.read_ident();
        // Read list of all params
        let params = self.read_param_list();
        // Everything from here until the closing bracket is part of the
        // component definition
        let body = self.read_component_body();
        let decl = ComponentDefinition::new(name, params, body);
        Node::ComponentDefinition(decl)
    }

    // A component is composed of a set of expressions/statements.
    fn read_component_body(&mut self) -> Vec<Expression> {
        let mut exprs = vec![];
        loop {
            match self.lexer.next_token() {
                // Statements starting with '<' are assumed to be
                // the beginning of JSX statements.
                Some(Token::LessThan) => {
                    let expr = Expression::JSXElement(self.read_jsx_element());
                    exprs.push(expr);
                    return exprs;
                }
                _ => {
                    continue;
                }
            }
        }
    }

    /**
     * read_expr
     * Reads any expresison, which is something that can evaluate to some value
     */
    fn read_expr(&mut self) -> Expression {
        self.skip_until(|token| token == Token::RCurlyBracket);
        Expression::NumericLiteral(42)
    }

    fn read_jsx_attributes(&mut self) -> Option<Vec<JSXAttribute>> {
        let mut attrs: Vec<JSXAttribute> = vec![];
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let attr = self.read_jsx_attribute();
                    attrs.push(attr);
                }
                Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                    break;
                }
                _ => {
                    unreachable!("Unknown token: {:?}", self.peek());
                }
            }
        }
        println!("read_jsx_attributes: {:?}", attrs);
        if attrs.is_empty() {
            None
        } else {
            Some(attrs)
        }
    }

    fn is_jsx_element_self_closing(&mut self) -> bool {
        match self.peek() {
            Some(&Token::GreaterThan) => {
                self.skip_assert(Token::GreaterThan);
                false
            }
            Some(&Token::ForwardSlash) => {
                self.skip_assert(Token::ForwardSlash);
                self.skip_assert(Token::GreaterThan);
                true
            }
            _ => {
                unreachable!();
            }
        }
    }

    fn read_jsx_element(&mut self) -> JSXElement {
        let tag_name = self.read_ident();
        let attributes = self.read_jsx_attributes();
        let self_closing = self.is_jsx_element_self_closing();
        let opening_element = JSXOpeningElement::new(tag_name.clone(), attributes, self_closing);
        println!("opening_element {:?}", opening_element);
        let (children, closing_element) = if self_closing {
            (None, None)
        } else {
            let children = self.read_jsx_children();
            let closing_element = self.read_jsx_closing_element();
            println!("closing_element {:?}", closing_element);
            println!("children {:?}", children);
            (Some(children), Some(closing_element))
        };
        let element = JSXElement::new(opening_element, closing_element, children);
        println!("element: {:#?}", element);
        element
    }

    fn read_jsx_closing_element(&mut self) -> JSXClosingElement {
        // Closing tag
        // self.skip_assert(Token::LessThan);
        // self.skip_assert(Token::ForwardSlash);
        let closing_tag_name = self.read_ident();
        // TODO parsing error
        // assert_eq!(tag_name, closing_tag_name);
        self.skip_assert(Token::GreaterThan);
        // TODO actually parse the closing tag
        JSXClosingElement::new(closing_tag_name)
    }

    fn read_jsx_text(&mut self) -> Box<JSXChildren> {
        let mut text = String::new();
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let word = self.read_ident();
                    // TODO respect whitespace when lexing instead of this hack
                    text = text + " " + &word;
                }
                Some(&Token::LessThan) | Some(&Token::LCurlyBracket) => {
                    break;
                }
                _ => {
                    unreachable!();
                }
            }
        }
        println!("read_jsx_text: {:?}", text);
        Box::new(JSXChildren::JSXText(text))
    }

    fn read_jsx_children(&mut self) -> Vec<Box<JSXChildren>> {
        let mut children: Vec<Box<JSXChildren>> = vec![];
        loop {
            match self.peek() {
                // JSXText
                Some(&Token::Ident(_)) => {
                    children.push(self.read_jsx_text());
                },
                Some(&Token::LessThan) => {
                    // A '<' token could be the start of another element
                    // or the closing element for the enclosing element
                    self.skip_assert(Token::LessThan);
                    match self.peek() {
                        Some(&Token::Ident(_)) => {
                            // An identifier means the start of a new element
                            let element = self.read_jsx_element();
                            children.push(Box::new(JSXChildren::JSXElement(element)));
                        }
                        Some(&Token::ForwardSlash) => {
                            self.skip_assert(Token::ForwardSlash);
                            return children;
                            // A '/' indicates a closing element
                        }
                        _ => {
                            unreachable!();
                        }
                    }
                }
                _ => break,
            }
        }
        children
        // let child = match token {
        //     // TODO JSX parsers respect the whitespace inside tags. Need to output tokens
        //     Some(Token::Ident(word)) => {
        //         JSXChildren::JSXText(word)
        //     }
        //     Some(Token::LessThan) => {
        //         let element = self.read_jsx_element();
        //         JSXChildren::JSXElement(element)
        //     }
        //     // TODO handle JSXExpressionContainer
        //     token @ _ => {
        //         unreachable!("Unexpected token: {:?}", token);
        //     }
        // };
        // Box::new(child)
    }

    /**
     * Reads the expression part of a JSX attribute.
     */
    fn read_jsx_attribute_value(&mut self) -> JSXAttributeValue {
        // An explicit assignment. The assigned value can
        // either be a string literal or an identifier in curly brackets
        match self.lexer.next_token() {
            Some(Token::String(literal)) => JSXAttributeValue::StringLiteral(literal),
            Some(Token::LCurlyBracket) => {
                let expr = self.read_expr();
                JSXAttributeValue::Expression(expr)
            }
            token @ _ => unreachable!("Unknown token: {:?}", token),
        }
    }

    /**
     * read_jsx_attr
     *
     * A JSX Attribute comes in two forms:
     *
     * 1. An explicit attribute where the value is passed
     *    in inline (<div id="1" />)
     * 2. An implicit attribute where we assume the attribute
     *    name is a reference to some identifier
     *    (<div id>)
     */
    fn read_jsx_attribute(&mut self) -> JSXAttribute {
        let name = self.read_ident();

        let foo = match self.peek() {
            Some(&Token::Assign) => {
                self.skip_assert(Token::Assign);
                let attribute_value = self.read_jsx_attribute_value();
                JSXAttribute::new(name, attribute_value)
            }
            Some(&Token::Ident(_)) | Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                let attr_ident = Identifier::new(name.clone());
                let attr_ident_expr = Expression::Identifer(attr_ident);
                let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
                JSXAttribute::new(name, attr_value)
            }
            _ => {
                unreachable!();
            }
        };
        foo
        // println!("read_jsx_attribute:name {:?}", name);
        // match self.lexer.next_token() {
        //     // Explicit assignment
        //     Some(Token::Assign) => {
        //         let attr_value = self.read_jsx_attribute_value();
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     // Shorthand form
        //     Some(Token::Ident(_)) | Some(Token::ForwardSlash) | Some(Token::GreaterThan) => {
        //         println!("shorthand form, creating the identifier and stuff");
        //         let attr_ident = Identifier::new(name.clone());
        //         let attr_ident_expr = Expression::Identifer(attr_ident);
        //         let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     token @ _ => unreachable!("Unknown character: {:?}", token),
        // }
    }

    fn read_type_def(&mut self) -> Node {
        // Next token is the identifier for the type definition.
        let name = self.read_ident();
        let def = TypeDefinition::new(name);
        // Just read everything until we reach the end for now...
        loop {
            if self.lexer.next_token() == Some(Token::RCurlyBracket) {
                break;
            }
        }
        Node::TypeDefinition(def)
    }

    /**
     * Inside a module, top level nodes are currently restricted to component
     * and function definitions. This means you cannot have other expressions
     * that are executed when the module is evalulated. All components and
     * functions must be lazily evaluated.
     */
    pub fn next_node(&mut self) -> Node {
        let token = self.lexer.next_token().unwrap();
        match token {
            Token::Reserved(ReservedWord::Component) => self.read_component_declaration(),
            Token::Reserved(ReservedWord::Type) => self.read_type_def(),
            _ => Node::Text,
        }
    }
}
