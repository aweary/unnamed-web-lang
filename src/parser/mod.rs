mod decl;
mod expr;
mod stmt;
mod context;

use decl::DeclParser;
// use expr::ExprParser;
// use stmt::StmtParser;
pub use context::{ParsingContext, ExprContext};


use crate::error::ParseError;
use crate::lexer::Lexer;
use crate::result::Result;
use crate::ast::{Module, Program};
use crate::symbol::Symbol;
use crate::token::{Token, TokenKind};
use std::rc::Rc;
use std::cell::RefCell;

// temporary use so that the Rust compiler checks this module
// use crate::ir::*;

pub fn parse_program<'a>(source: &'a str, mut ctx: &'a mut ParsingContext) -> Result<Program> {
    let mut parser = Parser::new(&source, &mut ctx);
    parser.parse()
}

pub struct Parser<'a> {
    lexer: Lexer<'a>,
    ctx: &'a mut ParsingContext,
}

impl<'a> Parser<'a> {
    pub fn new(source: &'a str, ctx: &'a mut ParsingContext) -> Self {
        let lexer = Lexer::new(&source);
        Parser { lexer, ctx }
    }

    fn next_token(&mut self) -> Result<Token> {
        match self.lexer.next_token(&mut self.ctx) {
            Ok(token) => {
                Ok(token)
            }
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn peek_token(&mut self) -> Result<&Token> {
        match self.lexer.peek_token(&mut self.ctx) {
            Ok(token) => Ok(token),
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn eat(&mut self, kind: TokenKind) -> Result<bool> {
        if self.peek_token()?.kind == kind {
            self.expect(kind)?;
            Ok(true)
        } else {
            Ok(false)
        }
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.next_token()?;
        if token.kind != kind {
            Err(ParseError::UnexpectedToken(token))
        } else {
            Ok(token)
        }
    }

    fn ident(&mut self) -> Result<Symbol> {
        let token = self.next_token()?;
        match token.kind {
            TokenKind::Ident(sym) => Ok(sym),
            _ => Err(ParseError::UnexpectedToken(token)),
        }
    }

    /**
     * This is currently an alias to `ident`, but a type
     * identifier has different syntax. For example, polymorphic
     * types have type parameters.
     */
    fn type_ident(&mut self) -> Result<Symbol> {
        self.ident()
    }

    pub fn parse(&mut self) -> Result<Program> {
        // top-level statements
        let stmts = self.decl_list()?;
        let module = Module { stmts };
        let modules = vec![module];
        let program = Program { modules };
        Ok(program)
    }
}
