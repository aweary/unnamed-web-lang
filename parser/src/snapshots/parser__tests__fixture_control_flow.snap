---
source: parser/src/lib.rs
expression: parse_module(input)
---
{
  "items": [
    {
      "kind": {
        "Fn": {
          "name": {
            "name": {
              "unstable_source": "control_flow"
            },
            "span": {
              "start": 3,
              "end": 15
            }
          },
          "params": {
            "Multi": [
              {
                "local": {
                  "Ident": [
                    {
                      "name": {
                        "unstable_source": "a"
                      },
                      "span": {
                        "start": 16,
                        "end": 17
                      }
                    },
                    {
                      "start": 16,
                      "end": 17
                    }
                  ]
                },
                "ty": "Existential",
                "id": 0,
                "span": {
                  "start": 16,
                  "end": 17
                }
              },
              {
                "local": {
                  "Ident": [
                    {
                      "name": {
                        "unstable_source": "b"
                      },
                      "span": {
                        "start": 19,
                        "end": 20
                      }
                    },
                    {
                      "start": 19,
                      "end": 20
                    }
                  ]
                },
                "ty": "Existential",
                "id": 0,
                "span": {
                  "start": 19,
                  "end": 20
                }
              }
            ]
          },
          "body": {
            "stmts": [
              {
                "id": 0,
                "kind": {
                  "If": {
                    "span": {
                      "start": 28,
                      "end": 103
                    },
                    "condition": {
                      "id": 0,
                      "kind": {
                        "Lit": {
                          "span": {
                            "start": 31,
                            "end": 35
                          },
                          "kind": {
                            "Bool": {
                              "unstable_source": "true"
                            }
                          }
                        }
                      },
                      "span": {
                        "start": 31,
                        "end": 35
                      },
                      "ty": null
                    },
                    "block": {
                      "stmts": [
                        {
                          "id": 0,
                          "kind": {
                            "Local": {
                              "id": 0,
                              "name": {
                                "Ident": [
                                  {
                                    "name": {
                                      "unstable_source": "a"
                                    },
                                    "span": {
                                      "start": 50,
                                      "end": 51
                                    }
                                  },
                                  {
                                    "start": 50,
                                    "end": 51
                                  }
                                ]
                              },
                              "ty": null,
                              "init": {
                                "id": 0,
                                "kind": {
                                  "Reference": {
                                    "name": {
                                      "unstable_source": "b"
                                    },
                                    "span": {
                                      "start": 54,
                                      "end": 55
                                    }
                                  }
                                },
                                "span": {
                                  "start": 54,
                                  "end": 55
                                },
                                "ty": null
                              },
                              "span": {
                                "start": 46,
                                "end": 55
                              }
                            }
                          },
                          "span": {
                            "start": 46,
                            "end": 56
                          },
                          "has_semi": false
                        }
                      ],
                      "id": 0,
                      "span": {
                        "start": 36,
                        "end": 62
                      }
                    },
                    "alt": {
                      "If": {
                        "span": {
                          "start": 68,
                          "end": 103
                        },
                        "condition": {
                          "id": 0,
                          "kind": {
                            "Lit": {
                              "span": {
                                "start": 71,
                                "end": 76
                              },
                              "kind": {
                                "Bool": {
                                  "unstable_source": "false"
                                }
                              }
                            }
                          },
                          "span": {
                            "start": 71,
                            "end": 76
                          },
                          "ty": null
                        },
                        "block": {
                          "stmts": [
                            {
                              "id": 0,
                              "kind": {
                                "Local": {
                                  "id": 0,
                                  "name": {
                                    "Ident": [
                                      {
                                        "name": {
                                          "unstable_source": "c"
                                        },
                                        "span": {
                                          "start": 91,
                                          "end": 92
                                        }
                                      },
                                      {
                                        "start": 91,
                                        "end": 92
                                      }
                                    ]
                                  },
                                  "ty": null,
                                  "init": {
                                    "id": 0,
                                    "kind": {
                                      "Lit": {
                                        "span": {
                                          "start": 95,
                                          "end": 96
                                        },
                                        "kind": {
                                          "Number": {
                                            "unstable_source": "4"
                                          }
                                        }
                                      }
                                    },
                                    "span": {
                                      "start": 95,
                                      "end": 96
                                    },
                                    "ty": null
                                  },
                                  "span": {
                                    "start": 87,
                                    "end": 96
                                  }
                                }
                              },
                              "span": {
                                "start": 87,
                                "end": 97
                              },
                              "has_semi": false
                            }
                          ],
                          "id": 0,
                          "span": {
                            "start": 77,
                            "end": 103
                          }
                        },
                        "alt": null
                      }
                    }
                  }
                },
                "span": {
                  "start": 28,
                  "end": 103
                },
                "has_semi": false
              },
              {
                "id": 0,
                "kind": {
                  "If": {
                    "span": {
                      "start": 108,
                      "end": 174
                    },
                    "condition": {
                      "id": 0,
                      "kind": {
                        "Binary": [
                          "DblEquals",
                          {
                            "id": 0,
                            "kind": {
                              "Reference": {
                                "name": {
                                  "unstable_source": "a"
                                },
                                "span": {
                                  "start": 111,
                                  "end": 112
                                }
                              }
                            },
                            "span": {
                              "start": 111,
                              "end": 112
                            },
                            "ty": null
                          },
                          {
                            "id": 0,
                            "kind": {
                              "Reference": {
                                "name": {
                                  "unstable_source": "b"
                                },
                                "span": {
                                  "start": 116,
                                  "end": 117
                                }
                              }
                            },
                            "span": {
                              "start": 116,
                              "end": 117
                            },
                            "ty": null
                          }
                        ]
                      },
                      "span": {
                        "start": 111,
                        "end": 117
                      },
                      "ty": null
                    },
                    "block": {
                      "stmts": [
                        {
                          "id": 0,
                          "kind": {
                            "Return": {
                              "id": 0,
                              "kind": {
                                "Reference": {
                                  "name": {
                                    "unstable_source": "a"
                                  },
                                  "span": {
                                    "start": 135,
                                    "end": 136
                                  }
                                }
                              },
                              "span": {
                                "start": 135,
                                "end": 136
                              },
                              "ty": null
                            }
                          },
                          "span": {
                            "start": 128,
                            "end": 137
                          },
                          "has_semi": false
                        }
                      ],
                      "id": 0,
                      "span": {
                        "start": 118,
                        "end": 143
                      }
                    },
                    "alt": {
                      "Block": {
                        "stmts": [
                          {
                            "id": 0,
                            "kind": {
                              "Return": {
                                "id": 0,
                                "kind": {
                                  "Reference": {
                                    "name": {
                                      "unstable_source": "b"
                                    },
                                    "span": {
                                      "start": 166,
                                      "end": 167
                                    }
                                  }
                                },
                                "span": {
                                  "start": 166,
                                  "end": 167
                                },
                                "ty": null
                              }
                            },
                            "span": {
                              "start": 159,
                              "end": 168
                            },
                            "has_semi": false
                          }
                        ],
                        "id": 0,
                        "span": {
                          "start": 149,
                          "end": 174
                        }
                      }
                    }
                  }
                },
                "span": {
                  "start": 108,
                  "end": 174
                },
                "has_semi": false
              },
              {
                "id": 0,
                "kind": {
                  "Return": {
                    "id": 0,
                    "kind": {
                      "Lit": {
                        "span": {
                          "start": 186,
                          "end": 187
                        },
                        "kind": {
                          "Number": {
                            "unstable_source": "4"
                          }
                        }
                      }
                    },
                    "span": {
                      "start": 186,
                      "end": 187
                    },
                    "ty": null
                  }
                },
                "span": {
                  "start": 179,
                  "end": 188
                },
                "has_semi": false
              }
            ],
            "id": 0,
            "span": {
              "start": 22,
              "end": 190
            }
          },
          "return_ty": "Existential",
          "is_async": false,
          "generics": null,
          "span": {
            "start": 0,
            "end": 190
          }
        }
      },
      "span": {
        "start": 0,
        "end": 190
      }
    }
  ]
}
